### ISSUE TITLE
loss= nan on running nb5 & nb6 from Chapter6

### ISSUE THREAD

**Issue by Aya-S** [PacktPublishing/Hands-On-Computer-Vision-with-Tensorflow#14]
I have been trying to debug this for a few days to no avail.
I have been running the notebooks as they are.
I'm getting 
Epoch  0/90: loss =   nan; v-loss =   nan; acc = 0.389; v-acc = 0.366; mIoU = 0.020; v-mIoU = 0.019
as output from training and can't figure out what is the problem

**Comment by benjaminplanche** [comment]
Hello @Aya-S,

Thank you for the feedback!
If everything else is working fine (data processing, model instantiation, etc.), my first guess would be that recent changes in TensorFlow are breaking the custom loss function introduced in those notebooks.

I.e., under-the-hood changes have been recently made to the way some losses are handled by TensorFlow. It is possible that it is affecting the way our custom loss is computed in those notebooks (as our loss relies on `tf.losses.SparseCategoricalCrossentropy`).

My suggestion to check that hypothesis would be to try running those notebooks with the TensorFlow version mentioned in the book (`pip install tensorflow==2.0.0-alpha0`). Adapting the custom loss function for the newest version could then be an interesting exercise (currently, the custom function overrides `__call__(...)` in order to weight the loss, e.g., in a per-class manner -- this could be removed if the weighting is not needed, or this could be done in a more compliant way otherwise). 

Hoping this helps!

**Comment by EliotAndres** [comment]
I would also add that deep nets can always suffer from unstable loss. Techniques to circumvent this exist. For instance you can start with a very small learning rate for the first few epochs and then raise it.

**Comment by Aya-S** [comment]
> Hello @Aya-S,
> 
> Thank you for the feedback!
> If everything else is working fine (data processing, model instantiation, etc.), my first guess would be that recent changes in TensorFlow are breaking the custom loss function introduced in those notebooks.
> 
> I.e., under-the-hood changes have been recently made to the way some losses are handled by TensorFlow. It is possible that it is affecting the way our custom loss is computed in those notebooks (as our loss relies on `tf.losses.SparseCategoricalCrossentropy`).
> 
> My suggestion to check that hypothesis would be to try running those notebooks with the TensorFlow version mentioned in the book (`pip install tensorflow==2.0.0-alpha0`). Adapting the custom loss function for the newest version could then be an interesting exercise (currently, the custom function overrides `__call__(...)` in order to weight the loss, e.g., in a per-class manner -- this could be removed if the weighting is not needed, or this could be done in a more compliant way otherwise).
> 
> Hoping this helps!

I changed the TF version and indeed that did the trick, thank you for the tip! i didn't come across the TF version in the book.
I managed to run training with the full class list using your same code.
now I'm trying to run this flow as a binary classifier ( road, no road and hit a wall) .
I'm Guessing tf.losses.SparseCategoricalCrossentropy isn't tailored for this scenario.. but at least the loss is not nan :) !

Thank you for the input @benjaminplanche  and @EliotAndres !

**Comment by benjaminplanche** [comment]
Glad it helped!
If your task target is indeed binary (e.g., road vs not-road), you can indeed switch to a different loss like a binary cross-entropy (BCE).

I will assume your issue is solved.
All the best!
